### ISSUE TITLE
Can't get FKF to work

### ISSUE THREAD

**Issue by jzmazlish** [waternumbers/FKF#14]
Pasting below code for an attempt at using FKF, which is returning me NA for the log-likelihood and filtered state variables (and not changing the initial values when I run it through the optimizer). There are a fair amount of NA's in the data I am using, but also many observations. If you have any idea why things aren't working, I would really appreciate the help.

#Optimizer
b0 <- rep(1, 39) #initial values
b1 <- rep(1, 39) 
b2 <- rep(-1, 39) 
gamma1 <- 0.5 #AR coefficient on first element of state vector (>=0)
gamma2 <- 0.5 #AR coeff on second element of state vector
g <- rep(1, 39)
x <- rep(10, 39)
z <- rep(10, 39)
par1 <- c(b0,b1,b2,gamma1,gamma2,g,x,z)
fit <- optim(par1, objfnc, control=list(maxit=100000), lower=c(-10000, -10000, --10000, -1, -1, 0, 0, 0),upper = c(5000, 5000, 5000, 1, 1, 5000, 5000, 5000), method="L-BFGS-B")
print(fit$par)

#objective function

w <- matrix(kalman_allDates$Gov_Debt_v_GDP, nrow=116, byrow=TRUE)
v = matrix(kalman_allDates$GDP_change, nrow=116, byrow = TRUE) #g_it in Aguiar
y = matrix(kalman_allDates$Close, nrow=116, byrow = TRUE)
objfnc <- function(theta) {
  #Destructure theta
  b0 <- theta[1] 
  b1 <- theta[2] 
  b2 <- theta[3] 
  gamma1 <- theta[4] #AR coefficient on first element of state vector (>=0)
  gamma2 <- theta[5] #AR coeff on second element of state vector
  g <- theta[6] 
  x <- theta[7]
  z <- theta[8]
    
    
  #Compute the matrices for the measurement equation
  ct <- matrix(b0 + b1 * w + 100*b2 * v, nrow=39) 
  zt <- array(dim=c(39,2,1)) 
  zt[, 1,] <- x 
  zt[, 2,] <- z 
  zt
  GGt <- diag(39) 
  GGt <- g * GGt 

  
  #Compute the matrices for the state equation
  dt <- matrix(c(0,0), nrow=2)
  Tt <- array(dim=c(2,2,1))
  Tt[ 1, 1, 1] <- gamma1
  Tt[1, 2, 1] <- 0
  Tt[2, 1, 1] <- 0
  Tt[2, 2, 1] <- gamma2
  HHt <- array(dim=c(2,2,1))
  HHt[1,1,1] <- 0.01 #normalized innovation volatilities
  HHt[2,2,1] <- 0.01
  HHt[1,2,1] <- HHt[2,1,1] <- 0

  #Initial values of the state vector
  a0 <- c(1,1)
  
  #Variances of a0
  #Follows the example in the package
  P0 <- 1000*diag(2)

  
  #The data
  yt <- matrix(y, nrow=39) 

  #Run the Kalman filter
  library(FKF)
  kf <- fkf(a0, P0, dt, ct, Tt, zt, HHt, GGt, yt)
  
  #Return the negative of the loglikelihood
  return(-kf$logLik)
}

**Comment by waternumbers** [comment]
Hi,

Happy to look into your issue further if there is a bug in FKF. However I can't determine this since your example isn't reproducible - it contains no data and some basic R coding errors (e.g. defining the the objective function after the call to the optimisation routine). 
I've altered your code above into a script which is attached. It won't run since there is no data but can act as a basis for providing a reproducible example that generates the error.

Regards
Paul  
[example.zip](https://github.com/waternumbers/FKF/files/10002185/example.zip)

**Comment by jzmazlish** [comment]
[KalmanData.xlsx](https://github.com/waternumbers/FKF/files/10011580/KalmanData.xlsx)
Hi Paul,

Thank you for your response, and sorry about the issues. I am pretty new to this. I've attached here the data frame I am using with the attempt at the Kalman filter. If you title it "kalman_allDates" the example template you gave me should run, but I am still encountering the same issues with NA log-likelihood and state variables. Sorry if there is an easier way to make things reproducible that I am missing.

**Comment by waternumbers** [comment]
No trouble, thanks for the data. 

Running the script with the data revealed that the initial parameter set was inconsistent with the length of the limit vectors and the way it was unpacked in objfun. I fixed this in the attached, see the comments.

This still produced and `NA` likelihood. The reason for this is that `FKF` handles missing values in the observations `yt`, but your data also has missing values in `ct` - the constant term in the observation equation (see the help file for `fks`). Since the values of ct are known it is simple to adjust the code in objfun so that the call to `fkf` fits `yt-ct` with a zero constant term in the observation eqn. In this way the NA values are handled correctly.

Attached code runs for me - note I resaved your xls file as a csv which is included
[example_v2.zip](https://github.com/waternumbers/FKF/files/10014190/example_v2.zip)
